src/Compiler.Dynamic/SystemWebCompilation.cs

```diff
@@ -87,119 +77,124 @@ protected IDisposable MarkRecompile()
 
     protected void RemovePage(string path) => _compiled.Remove(path);
 
-    protected Task<T> CompilePageAsync(string path, CancellationToken token)
+    protected void CompileAllPages(CancellationToken token)
     {
-        if (_compiled.TryGetValue(path, out var result))
+        var aspxFiles = Files.GetFiles().Where(t => t.FullPath.EndsWith(".aspx"))
+            .Select(t => t.FullPath);
+
+        foreach (var parser in GetParsersToCompile(aspxFiles))
         {
-            _logger.LogDebug("Retrieving previously compiled page '{Page}'", path);
-            return result;
+            _compiled[parser.CurrentVirtualPath.Path] = InternalCompilePage(parser, token);
         }
-        else
+    }
+
+    /// <summary>
+    /// We need to identify the order for building. Dependencies of a page/control/etc can be retrieved
+    /// by using <see cref="TemplateParser.GetDependencyPaths"/>, but we want to ensure we only compile something if its
+    /// dependencies have already been compiled.
+    /// </summary>
+    private IEnumerable<BaseTemplateParser> GetParsersToCompile(IEnumerable<string> files)
+    {
```

Comment:
this looks fine.. feel free to merge it... the issue is parser.parse.. I will see if TypedReference could be used... my hunch is do we need to change in TemplateCodeTreeGenerator if we are sticking with typedreference.